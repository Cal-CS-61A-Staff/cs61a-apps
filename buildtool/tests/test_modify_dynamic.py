from harness import AddDep, Environment, Input, Sh, create_test_env


def test_modifying_dynamic_deps(snapshot):
    with create_test_env(snapshot) as env:
        env: Environment

        env.annotate("Create two files: f1 and f2. ")
        env.annotate(
            "We explicitly depend on f1, but rely on an Input() to depend on f2 based on f1"
        )
        f1 = env.new_file()
        f2 = env.new_file()
        rule = env.declare_rule(
            AddDep(f1),
            input_action := Input([f1, f2], [AddDep(f2)]),
            shell_action := Sh([f1, f2]),
        )
        env.build(rule)

        env.annotate(
            "Even though this rule has an Input(), we can still rebuild it entirely from caches"
        )
        env.build(rule)

        env.annotate("We now modify f2 so we also depend on a new file f3")
        env.annotate(
            "Now we have to rerun the dependency detection, since it's possible that something has changed"
        )
        f3 = env.new_file()
        env.update_file(f2)
        input_action.update_result([f1, f2, f3], [AddDep(f2), AddDep(f3)])
        env.build(rule)
