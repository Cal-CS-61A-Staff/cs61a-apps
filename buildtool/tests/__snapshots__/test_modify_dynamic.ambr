# name: test_modifying_dynamic_deps
  '
  STARTING TEST
  
  # Create two files: f1 and f2. 
  # We explicitly depend on f1, but rely on an Input() to depend on f2 based on f1
  CREATING FILE f1.txt with contents: "V0 (f1.txt)"
  CREATING FILE f2.txt with contents: "V0 (f2.txt)"
  CREATING rule RuleA
  
  BUILDING rule RuleA
  RUNNING run.py --mode=INPUT --id=0 in working directory
  RUNNING run.py --mode=INPUT --id=0 in scratch directory with files ['f1.txt', 'f2.txt']
  RUNNING run.py --mode=SH --id=1 in scratch directory with files ['f1.txt', 'f2.txt']
  COMPLETED BUILDING rule RuleA
  
  # We now modify f2 so we also depend on a new file f3
  CREATING FILE f3.txt with contents: "V0 (f3.txt)"
  UPDATING FILE f2.txt with contents: "V1 (f2.txt)"
  
  BUILDING rule RuleA
  RUNNING run.py --mode=INPUT --id=0 in working directory
  RUNNING run.py --mode=INPUT --id=0 in scratch directory with files ['f1.txt', 'f2.txt', 'f3.txt']
  RUNNING run.py --mode=SH --id=1 in scratch directory with files ['f1.txt', 'f2.txt', 'f3.txt']
  COMPLETED BUILDING rule RuleA
  
  
  '
---
